### PR TITLE
Custom emoji add/edit/delete support added

### PR DESCRIPTION
**Pull request by Daniel-Worrall**
Integration with current Emoji object too.

- [x] RSpec
- [x] Code
- [x] Safe to Merge

### CODE REVIEW

**Comment by meew0** [comment]
Why did you disable that RuboCop check? I think it's pretty important as we don't want to have two kinds of line endings in the files. Look into Git's [`core.autocrlf`](https://git-scm.com/book/en/v2/Customizing-Git-Git-Configuration#Formatting-and-Whitespace) which takes care of this automatically.

**Comment by Daniel-Worrall** [comment]
New issues involving similar code will be fixed when rewritten to split emoji events into 3 due to nature of the code. (Create,Update,Delete)

File: lib/discordrb/events/guilds.rb [hunk 1]
```diff
@@ -59,12 +59,48 @@ class ServerUpdateEventHandler < ServerEventHandler; end
   # Server is deleted
   # @see Discordrb::EventContainer#server_delete
   class ServerDeleteEvent < ServerEvent
-    # Overide init_server to account for the deleted server
+    # Override init_server to account for the deleted server
```
Comment (Daniel-Worrall) [inline lib/discordrb/events/guilds.rb]: Forgot to mention, but typo fix :^)

File: .rubocop.yml [hunk 2]
```diff
@@ -63,3 +63,5 @@ Style/NumericLiterals:
 Lint/Eval:
   Exclude:
     - examples/*
+EndOfLine:
+  Enabled: false
```
Comment (meew0) [inline .rubocop.yml]: Somehow, this got re-added

File: lib/discordrb/bot.rb [hunk 3]
```diff
@@ -1104,6 +1111,40 @@ def handle_dispatch(type, data)
...
+      when :GUILD_EMOJIS_UPDATE
+        server_id = data['guild_id'].to_i
+        server = @servers[server_id]
+        old_emoji_data = server.emoji.clone
+        update_guild_emoji(data)
+        new_emoji_data = server.emoji
+
+        create = new_emoji_data.select do |k, _|
+          old_emoji_data[k].nil?
+        end.keys
```
Comment (meew0) [inline lib/discordrb/bot.rb]: This can be shortened to `create = new_emoji_data.keys - old_emoji_data.keys`

File: lib/discordrb/bot.rb [hunk 4]
```diff
@@ -1104,6 +1111,40 @@ def handle_dispatch(type, data)
...
+        delete = old_emoji_data.select do |k, _|
+          new_emoji_data[k].nil?
+        end.keys
```
Comment (meew0) [inline lib/discordrb/bot.rb]: This can be shortened to `delete = old_emoji_data.keys - new_emoji_data.keys`

File: lib/discordrb/bot.rb [hunk 5]
```diff
@@ -1104,6 +1111,40 @@ def handle_dispatch(type, data)
...
+        update = data.select do |k, v|
+          v.name != new_emoji_data(k).name || v.roles != new_emoji_data.roles
+        end.keys
```
Comment (meew0) [inline lib/discordrb/bot.rb]: `new_emoji_data` is neither a function nor does it have a `roles` property (test your code please).

I also think `create`, `delete` and `update` should be renamed to `created_ids`, `deleted_ids` and `updated_ids` respectively.

File: lib/discordrb/container.rb [hunk 6]
```diff
@@ -306,6 +306,53 @@ def server_delete(attributes = {}, &block)
...
+
+    # This **event** is raised when an emoji is updated.
+    # @param attributes [Hash] The event's attributes.
+    # @option attributes [String, Integer, Server] :server Matches the server.
+    # @option attributes [String, Integer] :id Matches the id of the emoji.
+    # @option attributes [String] :name Matches the name of the emoji.
+    # @option attributes [String] :old_name Matches the name of the emoji before the update.
+    # @yield The block is executed when the event is raised.
+    # @yieldparam event [ServerEmojiUpdateEvent] The event that was raised.
+    # @return [ServerEmojiUpdateEventHandler] The event handler that was registered.
```
Comment (meew0) [inline lib/discordrb/container.rb]: Please decapitalise the first word in `@return` statements so it doesn't show up as "Returns The event handler [...]" (also applies to the other shorthand adders)

File: lib/discordrb/data.rb [hunk 7]
```diff
@@ -1939,6 +1939,20 @@ def initialize(data, bot, server)
       process_roles(data['roles']) if server
     end
 
+    # Edit the name of the emoji. Requires manage_emojis.
+    # @return [Integer] emoji id
+    def edit(name)
```
Comment (meew0) [inline lib/discordrb/data.rb]: I think this method would be better suited as `name=`

File: lib/discordrb/data.rb [hunk 8]
```diff
@@ -1939,6 +1939,20 @@ def initialize(data, bot, server)
       process_roles(data['roles']) if server
     end
 
+    # Edit the name of the emoji. Requires manage_emojis.
+    # @return [Integer] emoji id
```
Comment (meew0) [inline lib/discordrb/data.rb]: Is there any reason the ID is returned specifically here? (also with `delete`)

File: lib/discordrb/data.rb [hunk 9]
```diff
@@ -1939,6 +1939,20 @@ def initialize(data, bot, server)
       process_roles(data['roles']) if server
     end
 
+    # Edit the name of the emoji. Requires manage_emojis.
```
Comment (meew0) [inline lib/discordrb/data.rb]: "Edits the name [...]"

File: lib/discordrb/data.rb [hunk 10]
```diff
@@ -2500,6 +2514,35 @@ def afk_timeout=(afk_timeout)
...
+    # @param name [String] The name given to the new custom emoji.
+    # @param emoji [String, #read] A JPG file to be used as the avatar, either something readable (e. g. File Object) or as a data URL.
+    def add_emoji(name, emoji)
+      if emoji.respond_to? :read
+        # Set the file to binary mode if supported, so we don't get problems with Windows
+        emoji.binmode if emoji.respond_to?(:binmode)
+
+        emoji_string = 'data:image/jpg;base64,'
+        emoji_string += Base64.strict_encode64(emoji.read)
+        priv_add_emoji(emoji_string, name)
```
Comment (meew0) [inline lib/discordrb/data.rb]: The `priv_add_emoji` method can be replaced by simply calling `add_emoji` recursively with the string and doing the actual sending in the `else` part.

File: lib/discordrb/events/guilds.rb [hunk 11]
```diff
@@ -59,12 +59,130 @@ class ServerUpdateEventHandler < ServerEventHandler; end
...
     def init_server(data, bot)
       @server = Discordrb::Server.new(data, bot, false)
     end
   end
 
   # Event handler for {ServerDeleteEvent}
   class ServerDeleteEventHandler < ServerEventHandler; end
+
+  # Generic subclass for emoji events (create/update/delete)
+  class ServerEmojiEvent < ServerEvent; end
```
Comment (meew0) [inline lib/discordrb/events/guilds.rb]: Why does this class exist? Only `ServerEmojiUpdateEvent` inherits from it and that doesn't use any functionality specific to `ServerEvent`.

File: lib/discordrb/events/guilds.rb [hunk 12]
```diff
@@ -59,12 +59,130 @@ class ServerUpdateEventHandler < ServerEventHandler; end
...
+    end
+
+    def process_emoji(data)
+      @emoji = data['emojis'].map do |e|
+        @server.emoji(e.id)
+      end
+    end
+  end
+
+  #
```
Comment (meew0) [inline lib/discordrb/events/guilds.rb]: This comment should not be empty

File: lib/discordrb/permissions.rb [hunk 13]
```diff
@@ -35,7 +35,8 @@ class Permissions
       25 => :use_voice_activity,   # 33554432
       26 => :change_nickname,      # 67108864
       27 => :manage_nicknames,     # 134217728
-      28 => :manage_roles          # 268435456, also Manage Permissions
+      28 => :manage_roles,         # 268435456, also Manage Permissions # 29 536870912
```
Comment (meew0) [inline lib/discordrb/permissions.rb]: There's a newline missing in the comment here

Comment (Daniel-Worrall) [inline lib/discordrb/permissions.rb]: This makes rubocop unhappy :(

**Comment by Daniel-Worrall** [comment]
Even though this PR appears safe to merge, I would still like to take some time to add specs to ensure it works as intended and will remain to do so in the future.

**Comment by meew0** [comment]
For the RuboCop thing, I'd suggest just setting
```rb
# rubocop:disable Style/ExtraSpacing
```
on the file.

**Comment by Daniel-Worrall** [comment]
Ah. You meant in the file itself. I'm a rubocop noob. Let me know which you'd prefer.

**Comment by Daniel-Worrall** [comment]
Note: This fixes #284

**Comment by meew0** [comment]
I think it'd be cleaner if the comment is in the file itself, so you can quickly see it when reading the file, but it really doesn't matter that much.

**Comment by Daniel-Worrall** [comment]
Fixes #295

**Pull request merged**